lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java

```diff
@@ -8,81 +8,174 @@
-        chain.initCause(err);
-        root = root.getParentFile();
+        if (searchFromDirToTop(
+            e, new File(".").getAbsoluteFile(), "target", "rust", "parser-jni", name)) {
+          return;
+        }
+        throw new IllegalStateException("Cannot load parser from " + root, e);
       }
     }
-    return false;
-  }
 
-  private long stateUnlessClosed;
+    private static boolean searchFromDirToTop(Throwable chain, File root, String... names) {
+      while (root != null) {
+        var parser = root;
+        for (var e : names) {
+          parser = new File(parser, e);
+        }
+        try {
+          System.load(parser.getAbsolutePath());
+          return true;
+        } catch (LinkageError err) {
+          while (chain.getCause() != null) {
+            chain = chain.getCause();
+          }
+          chain.initCause(err);
+          root = root.getParentFile();
+        }
+      }
+      return false;
+    }
+
+    private long stateUnlessClosed;
+
+    private Worker(long stateIn) {
+      stateUnlessClosed = stateIn;
+    }
 
-  private Parser(long stateIn) {
-    stateUnlessClosed = stateIn;
+    public static Worker create() {
+      initializeLibraries();
+      var state = allocState();
+      return new Worker(state);
+    }
+
+    public long isIdentOrOperator(CharSequence input) {
+      byte[] inputBytes = input.toString().getBytes(StandardCharsets.UTF_8);
+      ByteBuffer inputBuf = ByteBuffer.allocateDirect(inputBytes.length);
+      inputBuf.put(inputBytes);
+
+      return Parser.isIdentOrOperator(inputBuf);
+    }
+
+    private long getState() {
+      if (stateUnlessClosed != 0) {
+        return stateUnlessClosed;
+      } else {
+        throw new IllegalStateException("Parser used after close()");
+      }
+    }
+
+    public ByteBuffer parseInputLazy(CharSequence input) {
+      var state = getState();
+      byte[] inputBytes = input.toString().getBytes(StandardCharsets.UTF_8);
+      ByteBuffer inputBuf = ByteBuffer.allocateDirect(inputBytes.length);
+      inputBuf.put(inputBytes);
+      return parseTreeLazy(state, inputBuf);
+    }
+
+    public Tree parse(CharSequence input) {
+      var state = getState();
+      byte[] inputBytes = input.toString().getBytes(StandardCharsets.UTF_8);
+      ByteBuffer inputBuf = ByteBuffer.allocateDirect(inputBytes.length);
+      inputBuf.put(inputBytes);
+      var serializedTree = parseTree(state, inputBuf);
+      var base = getLastInputBase(state);
+      var metadata = getMetadata(state);
+      serializedTree.order(ByteOrder.LITTLE_ENDIAN);
+      var message = new Message(serializedTree, input, base, metadata);
+      try {
+        return Tree.deserialize(message);
+      } catch (BufferUnderflowException | IllegalArgumentException e) {
+        LoggerFactory.getLogger(this.getClass())
+            .error("Unrecoverable parser failure for: {}", input, e);
+        throw e;
+      }
+    }
+
+    @Override
+    public void close() {
```

Comment:
I don't think there is anyone who can call `close()` now! We need to associated an instance of Worker with `ReferenceQueue` and clean it up when a `Thread` finishes or it is not used for a long time. Otherwise we'll be leaking Rust memory...